### ISSUE TITLE
lax.cond can mark an array as replicated when it is not, giving rise to silent errors with shard map

### ISSUE THREAD

**Issue by spraharsh** [jax-ml/jax#40823]
## Examples
This issue popped up when dealing with https://github.com/patrick-kidger/equinox/pull/1262. I will explain it with a contrived gradient example below with 2 CPU devices

```python
import jax
import jax.numpy as jnp
import numpy as np
from jax import lax
from jax.sharding import Mesh, PartitionSpec as P

mesh = Mesh(np.array(jax.devices()), ("i",))

@jax.shard_map(
    mesh=mesh, in_specs=P(), out_specs=P("i"), check_vma=True
)
def f(x):
    return lax.cond(
        lax.axis_index("i") == 0,
        lambda x: 2 * x,
        lambda x: 3 * x,
        x,
    )[None]

print(f(2.))
print(jax.jit(jax.grad(lambda x: f(x).sum()))(2.))
```

The correct expected gradient is 5, but we end up getting 4. 

For a less contrived example where not having that information is a setback, I reproduce my example in the comment here https://github.com/patrick-kidger/equinox/pull/1262

```python
x = [0, 1, 2, 3]   # device 1 has [0, 1] , device 2 has [2, 3], edited
condition = x[0] < 1.5 
pred = lax.cond(condition, lambda: True, lambda: False) # lax.cond loses information that the predicate is varying but the array is actually # device 1 has [True, True], device 2 has [False False]

result = eqx.error_if(x, pred, "negative", on_error="nan") # attempted sharding assumes predicate is not varying i.e distributed, therefore it will try to forego pmax and can only give wrong result without comms
```

## Timeline of relevant issues/PRs
I'll only keep the illustrative ones, that are relevant to changes in behavior/have unexpected behavior (I filtered the list I got from chatgpt)

1. https://github.com/jax-ml/jax/issues/24418 Issue similar to one of my examples, where we get an exception raise because of the lack of replication rules
2. https://github.com/jax-ml/jax/pull/24877 PR adding replication rules for cond
3. https://github.com/jax-ml/jax/commit/76271d638ad94f3df854054640f3b35161ee5be4 Here the relevant code is 

```python
  b0_vma = [o.vma for o in branches[0].out_avals]
  for branch in branches[1:]:
    b_vma = [o.vma for o in branch.out_avals]
    if b0_vma != b_vma:
      raise Exception("The branches of cond produced mismatched varying manual "
                      f"axes. Got {b0_vma} and {b_vma}. Please open an issue "
                      "at https://github.com/jax-ml/jax/issues, and as a "
                      "temporary workaround pass the check_rep=False argument "
                      "to shard_map")
```

Note that the code only checks the branches of the axes. If the predicate itself is differently satisfied on different processors, it will lead to a varying output. 

The original replicating code was later deleted in https://github.com/jax-ml/jax/pull/27929

## fix
imo if `pred` is varying in cond(pred, true_fun, false_fun) then the output should be varying as well, and should not raise an error

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.11.2
jaxlib: 0.11.2
numpy:  2.2.6
python: 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:50:00) [GCC 14.3.0]
device info: cpu-2, 2 local devices"
process_count: 1
platform: uname_result(system='Linux', node='praharsh-work', release='6.14.0-37-generic', version='#37-Ubuntu SMP PREEMPT_DYNAMIC Fri Nov 14 22:10:32 UTC 2025', machine='x86_64')
JAX_PLATFORMS=cpu
JAX_NUM_CPU_DEVICES=2

**Comment by spraharsh** [comment]
I would like to work on this, the quickest fix looks like it is to add a guard so that

`lax.cond(pred, lambda: True, lambda: False)` would raise an error if the predicate is varying implying the user needs to write

```
lax.cond(
    pred,
    lambda: lax.pcast(True, "i", to="varying"),
    lambda: lax.pcast(False, "i", to="varying"),
)
```
Which I think makes sense from a parallelism perspective (From a model perspective, we're dealing with one copy of [True, True, True True] or [False False False False] if `pred` is varying and multiple copies of True and False for each shard of `pred` if pred is replicated. The only clean way of not introducing differences between devices when there are not is to explicitly state that beforehand.

I think the label `varying` is a bit confusing in this regard, I feel like it means 'distributed' rather than whether it's varying or not.

The other version is to let do variance propagation where if 

if pred is varying , left_fun is replicated, right_fun is replicated, we get a varying result. (Although gradients through that look iffy)
